metric/instrument/sync.go

```diff
@@ -0,0 +1,159 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package instrument // import "go.opentelemetry.io/otel/metric/instrument"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/embedded"
+)
+
+// Counter is an instrument that records increasing incremental values.
+//
+// Warning: Methods may be added to this interface in minor releases. See
+// [go.opentelemetry.io/otel/metric] package documentation on API
+// implementation for information on how to set default behavior for
+// unimplemented methods.
+type Counter[N int64 | float64] interface {
+	embedded.Counter[N]
+
+	// Add records a change to the counter.
+	Add(ctx context.Context, incr N, attrs ...attribute.KeyValue)
+}
+
+// CounterConfig contains options for synchronous counter instruments that
+// records increasing incremental values.
+type CounterConfig[N int64 | float64] struct {
+	description string
+	unit        string
+}
```

Comment:
Do you need config per record type? A.k.a what about CounterConfig not having the generic type.

Reply:
The Observable configs need to be split on number type given the callback from `WithCallback` is defined by number. It was done for all the configurations to provide consistency, but also to ensure the forward compatibility with OTel if they decide in the future to add options that are defined by number.

Reply:
Not sure I understand the "but also to ensure the forward compatibility with OTel if they decide in the future to add options that are defined by number." unless the capability is exactly the same but the implementation differs by the number type (which I believe you can solve at the construction time) I am failing to understand this benefit, maybe an example would be good.

For `WithCallback` is only for async, and you can do a runtime error during construction if you want by keeping the WithCallback generic and return a non generic Option and validate the type during construction (`meter.NewFoo`).